### PR TITLE
Pulpcore 3.13 compatibility changes

### PR DESCRIPTION
**Pull request by dralley**
[noissue]

### CODE REVIEW

**Comment by dralley** [comment]
@quba42

**Comment by dralley** [comment]
These are all of the changes that I know about, but I didn't run the test suite to see if any warnings got printed.

I don't really have time to make the auto-publish and auto-distribute, changes, but the RPM plugin is a good reference.  The old PR isn't useful anymore because we changed course about the way it should be done.  Luckily the new way is much simpler.

**Review by mdellweg — COMMENTED** [review]
Codewise this looks good. Still should be tested with real data.

File: pulp_deb/app/models/repository.py [hunk 1]
```diff
@@ -55,3 +55,11 @@ def finalize_new_version(self, new_version):
...
         remove_duplicates(new_version)
         validate_repo_version(new_version)
+
+    def on_new_version(self, version):
+        """
+        Called when new repository versions are created.
+        Args:
+            version: The new repository version.
+        """
+        super().on_new_version(version)
```
[the remark this answers is not in the document]
Comment (quba42) [reply pulp_deb/app/models/repository.py:65]: Wouldn't it inherit that function from `super()` anyway, even without adding this code?
I feel like I am missing something basic and obvious here...

File: pulp_deb/app/serializers/publication_serializers.py [hunk 2]
```diff
@@ -51,11 +56,19 @@ class Meta:
...
     """
     Serializer for AptDistributions.
     """
 
+    publication = DetailRelatedField(
+        required=False,
+        help_text="Publication to be served",
+        view_name_pattern=r"publications(-.*/.*)?-detail",
+        queryset=Publication.objects.exclude(complete=False),
+        allow_null=True,
```
Comment (quba42) [inline pulp_deb/app/serializers/publication_serializers.py:69]: Does this mean I can create a distribution that does not serve anything?

File: pulp_deb/app/models/repository.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mdellweg) [reply pulp_deb/app/models/repository.py:65]: Yes, that it what i tried to say. You would extend this function to trigger auto publishing from every newly created version.

File: pulp_deb/app/serializers/publication_serializers.py (hunk 2 above)
Comment (mdellweg) [reply pulp_deb/app/serializers/publication_serializers.py:69]: As a placeholder? I believe yes. It still reserves the basepath on the content app.

File: pulp_deb/app/migrations/0014_swap_distribution_model.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    dependencies = [
+        ('core', '0062_add_new_distribution_mastermodel'),
+        ('deb', '0013_aptremote_ignore_missing_package_indices'),
+    ]
+
+    operations = [
+        migrations.CreateModel(
+            name='NewAptDistribution',
+            fields=[
+                ('distribution_ptr', models.OneToOneField(auto_created=True, on_delete=django.db.models.deletion.CASCADE, parent_link=True, primary_key=True, related_name='deb_aptdistribution', serialize=False, to='core.Distribution')),
```
Comment (quba42) [inline pulp_deb/app/migrations/0014_swap_distribution_model.py:59]: Break this line somehow, maybe? (I am surprised black/flake did not complain.)

File: pulp_deb/app/models/repository.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (quba42) [reply pulp_deb/app/models/repository.py:65]: Ahh, so these lines function a bit like a TODO comment.

File: pulp_deb/app/serializers/publication_serializers.py [hunk 4]
```diff
@@ -51,11 +56,19 @@ class Meta:
...
-class AptDistributionSerializer(PublicationDistributionSerializer):
+class AptDistributionSerializer(DistributionSerializer):
     """
     Serializer for AptDistributions.
     """
 
+    publication = DetailRelatedField(
+        required=False,
+        help_text="Publication to be served",
+        view_name_pattern=r"publications(-.*/.*)?-detail",
```
[the remark this answers is not in the document]
Comment (quba42) [reply pulp_deb/app/serializers/publication_serializers.py:67]: Is the `view_name` for publications somehow automatically built from the `TYPE` string? (I don't fully understand what this pattern is trying to match against.)

File: pulp_deb/app/migrations/0014_swap_distribution_model.py (hunk 3 above)
Comment (mdellweg) [reply pulp_deb/app/migrations/0014_swap_distribution_model.py:59]: As migrations are usually autogenerated, they are on the ignorelist of black.

File: pulp_deb/app/serializers/publication_serializers.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mdellweg) [reply pulp_deb/app/serializers/publication_serializers.py:67]: It is built by the viewset. I think it is `<Mastermodel.endpoint_name>-<app_label>/<Model.endpoint_name>-<View.action>`.
So you could try: `r"publication-deb/(.*)-detail` here.

File: pulp_deb/app/models/repository.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mdellweg) [reply pulp_deb/app/models/repository.py:65]: I would still either remove or comment them.

[the remark this answers is not in the document]
Comment (dralley) [reply pulp_deb/app/models/repository.py:65]: Yes, it's basically just a placeholder.

File: pulp_deb/app/serializers/publication_serializers.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dralley) [reply pulp_deb/app/serializers/publication_serializers.py:67]: I value consistency across plugins more than being specific for this sort of thing.

File: pulp_deb/app/models/repository.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (quba42) [reply pulp_deb/app/models/repository.py:65]: I will look into adding the actual feature based on what I find in other plugins.

**Comment by quba42** [comment]
@dralley Thanks for providing this. If you don't mind, I am happy to take over this branch going forward.

I want to run some manual testing of the migration (as suggested by @mdellweg) and also add the auto publish feature.
Hopefully I can get that all done soon.

**Review by quba42 — CHANGES_REQUESTED** [review]
I have now tested this change for existing distributions migrating both up and back down, without apparent issues.

@dralley Can you still add the (cosmetic) suggestions from https://github.com/quba42/pulp_deb/commit/29c844bec16bccd818469b39fcde4927655df17a to your commit, then I will merge this (and work on adding the auto publish feature on a new PR on top).

**Comment by dralley** [comment]
@quba42 done

**Review by quba42 — APPROVED** [review]

**Pull request merged**